### ISSUE TITLE
Python msolve_selected can return same solution on repeated calls

### ISSUE THREAD

**Issue by ozppupbg** [msoos/cryptominisat#687]
Hello,

I just noticed, that `msolve_selected` in the Python API can return the same solution multiple times.
When I checked the source-code I found the reason here:
https://github.com/msoos/cryptominisat/blob/0bc084f15cb3ef94618817372d6cba29a86868c4/python/src/pycryptosat.cpp#L937-L977
As you can see, `current_nr_of_solutions` is increased before the if and the current solution is only banned, if the maximum has not yet been reached - which is not the case for the last found solution!

This bahaviour contradicts my intuition how this function should behave and also the documentation - as far as I interpret it:
If I have some clauses, which have 16 possible solutions and I call `msolve_selected` with `max_nr_of_solutions=16`, I would expect a following call to `solve` or `msolve_selected` to return UNSAT. But currently, it returns the last solution found in the first `msolve_selected` call.

Thank you!

**Comment by msoos** [comment]
Oh wow, that's almost certainly a bug. Do you have a fix for it? Wanna create a PR to fix? I can also fix, but I'd be happy to merge a PR you have :)

Mate

**Comment by msoos** [comment]
(Also, thanks a lot for finding it!!)

**Comment by ozppupbg** [comment]
I can create a PR, but it will take some time - I don't have a Dev environment set up at the moment.
I'm planning to do that anyway to test some other stuff, but it might take a couple of weeks.

**Comment by ozppupbg** [comment]
Currently I see three variants, how to proceed with the `var_selected` logic:

1. Leave it as is: If `var_selected` is empty, an empty clause is added and the problem becomes unsatisfiable
2. Raise an error if `var_selected` is empty
3. Do not add a clause if `var_selected` is empty and return the same solution multiple times, if `msolve_selected` is called repeatedly with empty `var_selected`

I'm currently unsure if 2 or 3 is my favorite solution.
How do you see this?

**Comment by msoos** [comment]
I have now fixed this. The `var_selected` always needs to be given. It's basically the projection set. If you give [1,2,3], it will give solutions over variables 1..3 only, so at most 8 solutions. If you give all variables in the CNF, it will iterate through all solutions to the CNF. This is the expected behaviour, because that's what `c ind ... 0` in the CNF  + running the binary with `--maxsol 1000000000` will do.

Please let me know what you think. Cheers,

Mate

**Comment by ozppupbg** [comment]
I'm glad, I could inspire you for another fix, but this issue was not, what my last question was about ;)

> Currently I see three variants, how to proceed with the `var_selected` logic:
> 
>     1. Leave it as is: If `var_selected` is empty, an empty clause is added and the problem becomes unsatisfiable
> 
>     2. Raise an error if `var_selected` is empty
> 
>     3. Do not add a clause if `var_selected` is empty and return the same solution multiple times, if `msolve_selected` is called repeatedly with empty `var_selected`
> 
> 
> I'm currently unsure if 2 or 3 is my favorite solution. How do you see this?

My question was about the case that `var_lits.size()==0` and thus `ban_solution` is an empty vector. This happens, if `msolve_selected` is called with a empty list as the `var_selected` parameter.

**Comment by msoos** [comment]
Ah, true. Sorry. Yeah, I think (2) is the best option for the moement. We could actually return 0/1, i.e. UNSAT/SAT in case it's an empty projection set. I will ask my collaborator, Kuldeep Meel, what he thinks. I think the UNSAT/SAT interpretation for empty sets is the right one, And then when it's over 1 variable, then we can return UNSAT/1/2, when it's over 2, it's UNSAT/1/2/3, etc. Because then we have 0..1+2^k every time as the potential return value. With k=0, we have 2, with k=1, we have 3, with k=2 we have 5, etc.

**Comment by kuldeepmeel** [comment]
Just chiming in: 

TL;DR: I agree with Mate's viewpoint. 

So far, our interpretation has been that we have a formula F(X) and then if we set projection set to P, then we are essentially asking for solutions of \exists (X \setminus P) F(X). 

Now if P is empty, then we have X\setminus P = X. Therefore, \exists X F(X) is either TRUE or FALSE. From this point of view, I think returning SAT or UNSAT is the right interpretation.

**Comment by msoos** [comment]
I have decided to delete `msolve_selected`. It is **always** possible to simply ban the solution in Python. This is much-much more user-friendly, and the performance overhead is minimal. Basically:
1. call `solve()`
2. call `add_clause()` to ban the part of the solution (e.g. the variables over the independent set)
3. goto 1.

This is explained how to do in the README: https://github.com/msoos/cryptominisat#incremental-python-usage

This closes this issue, since the problem is non-existent now. `msolve_selected` can now be trivially simulated from Python itself. This is more convenient, and provides higher level of control to the user.

Mate
